kythe/cxx/common/kzip_writer.cc

```diff
@@ -112,19 +125,39 @@ StatusOr<std::string> KzipWriter::WriteUnit(
     }
     initialized_ = true;
   }
-  if (auto json = WriteMessageAsJsonToString(unit)) {
-    auto file = InsertFile(kUnitRoot, *json);
+  auto json = WriteMessageAsJsonToString(unit);
+  if (!json) {
+    return json.status();
+  }
+  auto digest = SHA256Digest(*json);
+  StatusOr<std::string> result(InternalError("unsupported encoding"));
+  if (encoding_ == KzipEncoding::Json || encoding_ == KzipEncoding::All) {
+    auto file = InsertFile(absl::StrCat(kJsonUnitRoot, digest), *json);
     if (file.inserted()) {
       auto status = WriteTextFile(archive_, file.path(), file.contents());
       if (!status.ok()) {
         contents_.erase(file.path());
         return status;
       }
```

Comment:
minor: The `if (file.inserted())` logic should be shared, likely moved to `InsertFile` (unless I'm missing calls to InsertFile which aren't immediately followed by the same logic, but they all appear to be).

It was only used in two places before, so largely unnecessary, but now there's the magic number of > 2 ;-)

I think that refactoring will also allow us to get rid of InsertionResult entirely, as all of the existing uses immediately call WriteTextFile if the insertion succeeded and then return the digest (or error).

Reply:
Great idea.